### PR TITLE
[4/n][trainer] feat: flowgrpo - add diffusers + fsdp engine support

### PR DESCRIPTION
**Pull request by zhtmike**
### What does this PR do?

- Add Diffusers with FSDP as the training engine, for diffusion model RL.
- Add FlowGRPO algorithm (loss only, for UT testing)
- SP support with be provided in the separeted PR
- It might be too big to have trainer + engine PR together, so we split it into two. The next PR will be the final trainer PR, and the last one is doc PR for FlowGRPO algorithm.

Based on the PR: #5297
Relavant PRs: #5716 #5713 #5616 
~~Wait for #5716 to review first since there are some repeated codes~~

> Add **concise** overview of what this PR aims to achieve or accomplish. Reference related GitHub issues and PRs that help with the review.

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [x] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/trainer/ppo/core_algos.py [hunk 1]
```diff
@@ -2064,6 +2065,61 @@ def compute_policy_loss_cispo(
     return pg_loss, pg_metrics
 
 
+@register_policy_loss("flow_grpo")
```
Comment (wuxibin89) [inline verl/trainer/ppo/core_algos.py]: How about move diffusion related loss to a separate file? The core_algos.py is a bit too long.

Comment (zhtmike) [reply verl/trainer/ppo/core_algos.py]: Agree. Moved to a separated `diffusion_algos.py‎`

File: verl/utils/fsdp_utils.py [hunk 2]
```diff
@@ -590,7 +590,7 @@ def fsdp2_clip_grad_norm_(parameters, max_norm, norm_type=2.0, error_if_nonfinit
     return total_norm
 
 
-def layered_summon_lora_params(fsdp_module) -> OrderedDict:
+def layered_summon_lora_params(fsdp_module, is_diffusers=False) -> OrderedDict:
```
Comment (SamitHuang) [inline verl/utils/fsdp_utils.py:593]: considering future adaptation on veomni, should we use some var like`diffusion_backend = "diffusers"` instead?

File: verl/workers/engine/fsdp/diffusers_impl.py [hunk 3]
```diff
@@ -0,0 +1,839 @@
...
+
+        if peft_config is not None and base_sync_done:
+            per_tensor_param = params.items()
+        else:
+            device = get_device_id()  # used when fsdp2 set cpu_offload_policy
+            # TODO: cast fp32 to bf16 to reduce weight sync overhead, need more fine-grained control, e.g MoE gate
+            per_tensor_param = (
+                (
+                    name,
+                    param.to(device, non_blocking=True).full_tensor().to(torch.bfloat16, non_blocking=True)
```
Comment (SamitHuang) [inline verl/workers/engine/fsdp/diffusers_impl.py:768]: Ideally, we shouldcovert weights in mixed precision (like LayerNorm keep fp32) ? Does original flow_grpo repo also convert all weights to bf16 for weight sync?

**Review by SamitHuang — COMMENTED** [review]
can supplement the e2e running script in examples and attach the running results in the PR description

File: verl/utils/fsdp_utils.py (hunk 2 above)
Comment (zhtmike) [reply verl/utils/fsdp_utils.py:593]: this is for FSDP engine only. Only two backends can use it now (transformers or diffusers)

File: verl/workers/engine/fsdp/diffusers_impl.py (hunk 3 above)
Comment (zhtmike) [reply verl/workers/engine/fsdp/diffusers_impl.py:768]: > Ideally, we shouldcovert weights in mixed precision (like LayerNorm keep fp32) ?

Agree, but currently we follow Transformer + FSDP style. Need to refactor them together.

> Does original flow_grpo repo also convert all weights to bf16 for weight sync?

And there is no weight sync in flow_grpo repo.

**Comment by zhtmike** [comment]
> can supplement the e2e running script in examples and attach the running results in the PR description

It might be too large for a PR with e2e style training script, contains 40+ modified files, @AndyZhou952 am I right? 
It will be provided in the next PR.

File: verl/workers/engine/fsdp/diffusers_impl.py [hunk 4]
```diff
@@ -0,0 +1,839 @@
...
+                    if isinstance(param, DTensor)
+                    else param,
+                )
+                for name, param in params.items()
+            )
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply verl/workers/engine/fsdp/diffusers_impl.py:773]: It is a DTensor for FSDP1 with device mesh provided. So it is fine.

**Comment by AndyZhou952** [comment]
> > can supplement the e2e running script in examples and attach the running results in the PR description
> 
> It might be too large for a PR with e2e style training script, contains 40+ modified files, @AndyZhou952 am I right? It will be provided in the next PR.

There will be roughly ~15 additional files. Those will be in the next PR `[5/n][trainer] ...` part w/ e2e scripts

File: verl/workers/engine/fsdp/diffusers_impl.py [hunk 5]
```diff
@@ -0,0 +1,839 @@
...
+        init_context = get_init_weight_context_manager(use_meta_tensor=True, mesh=self.device_mesh)
+
+        with init_context(), warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+
+            module = AutoModel.from_pretrained(
+                self.model_config.local_path,
+                torch_dtype=torch_dtype,
+                trust_remote_code=self.model_config.trust_remote_code,
+                subfolder="transformer",
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply verl/workers/engine/fsdp/diffusers_impl.py]: Supporting a UNet backbone carries too much historical burden, let’s stick with the transformer backbone only here.
Comment is added.

File: verl/workers/engine/fsdp/diffusers_impl.py [hunk 6]
```diff
@@ -0,0 +1,839 @@
...
+            # Forward and backward for each timestep
+            with ctx:
+                for step in range(micro_batch["all_timesteps"].shape[1]):
+                    loss, meta_info = self.forward_step(
+                        micro_batch, loss_function=loss_function, forward_only=forward_only, step=step
+                    )
+
+                    if not forward_only:
+                        loss.backward()
+
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply verl/workers/engine/fsdp/diffusers_impl.py]: resolved.

File: verl/workers/engine/fsdp/diffusers_impl.py [hunk 7]
```diff
@@ -0,0 +1,839 @@
...
+        if isinstance(negative_prompt_embeds, torch.Tensor) and negative_prompt_embeds.is_nested:
+            negative_prompt_embeds, negative_prompt_embeds_mask = self._unpad_nested_embeds(
+                negative_prompt_embeds, negative_prompt_embeds_mask
+            )
+
+        height = tu.get_non_tensor_data(data=micro_batch, key="height", default=None)
+        width = tu.get_non_tensor_data(data=micro_batch, key="width", default=None)
+        vae_scale_factor = tu.get_non_tensor_data(data=micro_batch, key="vae_scale_factor", default=None)
+        img_shapes = [[(1, height // vae_scale_factor // 2, width // vae_scale_factor // 2)]]
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply verl/workers/engine/fsdp/diffusers_impl.py]: nice suggestion. Refactored.

File: verl/workers/engine_workers.py [hunk 8]
```diff
@@ -113,8 +113,8 @@ def __init__(self, config: TrainingWorkerConfig):
 
         # we use the one defined in model
         # TODO: this is not elegant and should refactor later
-        self.engine_config.use_remove_padding = self.model_config.use_remove_padding
-        self.engine_config.use_fused_kernels = self.model_config.use_fused_kernels
+        self.engine_config.use_remove_padding = self.model_config.get("use_remove_padding", False)
+        self.engine_config.use_fused_kernels = self.model_config.get("use_fused_kernels", False)
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply verl/workers/engine_workers.py:117]: it is fine.

File: verl/workers/engine/fsdp/diffusers_impl.py [hunk 9]
```diff
@@ -0,0 +1,827 @@
...
+            torch_dtype = torch.float32 if not self.engine_config.forward_only else torch.bfloat16
+
+        torch_dtype = PrecisionType.to_dtype(torch_dtype)
+
+        init_context = get_init_weight_context_manager(use_meta_tensor=True, mesh=self.device_mesh)
+
+        with init_context(), warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+
+            module = AutoModel.from_pretrained(
```
Comment (wuxibin89) [inline verl/workers/engine/fsdp/diffusers_impl.py:183]: Do we support ulysses sequence parallel for DiT model?

Comment (zhtmike) [reply verl/workers/engine/fsdp/diffusers_impl.py:183]: Yes! but only the main branch of the diffusers supports. I will provide a PR of SP support once the next version of diffusers releases.

File: verl/workers/engine/fsdp/diffusers_impl.py [hunk 10]
```diff
@@ -0,0 +1,827 @@
...
+        output_lst = []
+
+        ctx = torch.no_grad() if forward_only else nullcontext()
+
+        for micro_batch in micro_batches:
+            micro_batch = micro_batch.to(get_device_id())
+            meta_info_lst = {"model_output": [], "loss": [], "metrics": []}
+            # Forward and backward for each timestep
+            with ctx:
+                for step in range(num_timesteps):
```
Comment (SamitHuang) [inline verl/workers/engine/fsdp/diffusers_impl.py:461]: can we use torch.compile to accelerate the loop for each time step?

File: verl/workers/engine/fsdp/diffusers_impl.py [hunk 11]
```diff
@@ -0,0 +1,827 @@
...
+        auto_wrap_policy = get_fsdp_wrap_policy(
+            module=module,
+            config=self.engine_config.wrap_policy,
+            is_lora=self.model_config.lora_rank > 0,
+        )
+
+        fsdp_mesh = self.device_mesh
+        sharding_strategy = get_sharding_strategy(fsdp_mesh)
+
+        # Note: We force turn off CPUOffload because it causes incorrect results when using grad accumulation
```
Comment (SamitHuang) [inline verl/workers/engine/fsdp/diffusers_impl.py:259]: what is the root cause for the incorrect result? FSDP + Offload can largely reduce  the memory cost

File: verl/workers/engine/fsdp/diffusers_impl.py [hunk 12]
```diff
@@ -0,0 +1,827 @@
...
+        init_context = get_init_weight_context_manager(use_meta_tensor=True, mesh=self.device_mesh)
+
+        with init_context(), warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+
+            module = AutoModel.from_pretrained(
+                self.model_config.local_path,
+                torch_dtype=torch_dtype,
+                trust_remote_code=self.model_config.trust_remote_code,
+                subfolder="transformer",  # currently we support DiT with transformer backbone only.
```
Comment (SamitHuang) [inline verl/workers/engine/fsdp/diffusers_impl.py:187]: There can be two transformers for some diffusion models , like Wan2.2. It’s better adapt for this case

File: verl/workers/engine/fsdp/diffusers_impl.py (hunk 10 above)
Comment (zhtmike) [reply verl/workers/engine/fsdp/diffusers_impl.py:461]: It is not so popular in the large scale training frameworks. But it is indeed a research question.

File: verl/workers/engine/fsdp/diffusers_impl.py (hunk 11 above)
Comment (zhtmike) [reply verl/workers/engine/fsdp/diffusers_impl.py:259]: It is directly copied from transformer + fsdp, to make them have the same training behaviour. Can dig into / fix them together.

File: verl/workers/engine/fsdp/diffusers_impl.py (hunk 12 above)
Comment (zhtmike) [reply verl/workers/engine/fsdp/diffusers_impl.py:187]: Agree. Can refactor it once new model with extra transformer (only wan2.2 now?) is added.

File: verl/workers/engine_workers.py [hunk 13]
```diff
@@ -144,7 +144,11 @@ def __init__(self, config: TrainingWorkerConfig):
             is_collect=self.engine.is_mp_src_rank_with_outputs(),
         )
 
-        self.flops_counter = FlopsCounter(self.model_config.hf_config)
+        if hasattr(self.model_config, "hf_config"):
+            self.flops_counter = FlopsCounter(self.model_config.hf_config)
+        else:
+            # for Diffusion models, FlopsCounter is not supported yet.
```
Comment (SamitHuang) [inline verl/workers/engine_workers.py:150]: Maybe we can add FlopsCounter for Diffusion Models in Q2?

Comment (zhtmike) [reply verl/workers/engine_workers.py:150]: Agree.

File: verl/workers/engine/fsdp/diffusers_impl.py [hunk 14]
```diff
@@ -0,0 +1,827 @@
...
+                "target_parameters": convert_to_regular_types(self.model_config.target_parameters),
+                "exclude_modules": convert_to_regular_types(self.model_config.exclude_modules),
+                "bias": "none",
+            }
+            module.add_adapter(LoraConfig(**lora_config))
+
+        return module
+
+    def _build_fsdp_module(self, module):
+        # TODO(ziheng): need to improve
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply verl/workers/engine/fsdp/diffusers_impl.py:233]: It is directly copied from transformer + fsdp, to make them have the same training behaviour. Can dig into / fix them together.

File: verl/workers/engine/fsdp/diffusers_impl.py [hunk 15]
```diff
@@ -0,0 +1,827 @@
...
+        log_gpu_memory_usage("Before offload_fsdp_model_to_cpu", logger=logger)
+        if self._is_offload_param:
+            offload_fsdp_model_to_cpu(self.module)
+        log_gpu_memory_usage("After offload_fsdp_model_to_cpu", logger=logger)
+
+        if peft_config is not None and base_sync_done:
+            per_tensor_param = params.items()
+        else:
+            device = get_device_id()  # used when fsdp2 set cpu_offload_policy
+            # TODO: cast fp32 to bf16 to reduce weight sync overhead, need more fine-grained control, e.g MoE gate
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply verl/workers/engine/fsdp/diffusers_impl.py:764]: same
